piet-direct2d/src/dwrite.rs

```diff
@@ -178,81 +268,113 @@ impl TextFormatBuilder {
     }
 }
 
-impl TextLayoutBuilder {
-    pub fn new(factory: DwriteFactory) -> TextLayoutBuilder {
-        TextLayoutBuilder {
-            factory,
-            format: None,
-            text: None,
-            width: None,
-            height: None,
-            alignment: TextAlignment::default(),
-        }
-    }
-
-    pub fn format(mut self, format: &TextFormat) -> TextLayoutBuilder {
-        // The fact we clone here is annoying, but it gets us out of
-        // otherwise annoying lifetime issues.
-        self.format = Some(format.clone());
-        self
-    }
-
-    pub fn text(mut self, text: &str) -> TextLayoutBuilder {
-        self.text = Some(text.to_wide());
-        self
+impl AsRef<str> for FamilyName {
+    fn as_ref(&self) -> &str {
+        &self.name
     }
+}
 
-    pub fn width(mut self, width: f32) -> TextLayoutBuilder {
-        self.width = Some(width);
-        self
+impl AsRef<[u16]> for FamilyName {
+    fn as_ref(&self) -> &[u16] {
+        &self.wide_name
     }
+}
 
-    pub fn height(mut self, height: f32) -> TextLayoutBuilder {
-        self.height = Some(height);
-        self
-    }
+#[allow(overflowing_literals)]
+#[allow(clippy::unreadable_literal)]
+const E_NOT_SUFFICIENT_BUFFER: HRESULT = 0x8007007A;
 
-    pub fn alignment(mut self, alignment: TextAlignment) -> TextLayoutBuilder {
-        self.alignment = alignment;
-        self
+fn make_text_range(start: usize, len: usize) -> DWRITE_TEXT_RANGE {
+    DWRITE_TEXT_RANGE {
+        startPosition: start as u32,
```

Comment:
The `as` conversions here really should be `try_into` with panicking unwrap. That said, unlikely to be an issue in practice.

Reply:
hm, how about a saturating cast? `start::try_into::<u32>().unwrap_or(u32::max_value())`?

Reply:
That can, among other things, cut a code point. I prefer panicking. Obviously we will never see a 4GB string for reals, it's invariably either an attack or a logic bug.